src/linear_algebra/dual.lean

```diff
@@ -398,3 +407,192 @@ begin
 end
 
 end dual_pair
+
+namespace submodule
+
+universes u v w
+
+variables {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M] [module R M]
+variable {W : submodule R M}
+
+/-- The `dual_annihilator` of a submodule `W` is the set of linear maps `φ` such
+  that `φ w = 0` for all `w ∈ W`. -/
+def dual_annihilator {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M]
+  [module R M] (W : submodule R M) : submodule R $ module.dual R M :=
+{ carrier := { φ | ∀ w ∈ W, φ w = 0 },
+  zero_mem' := by simp,
+  add_mem' :=  by { intros φ ψ hφ hψ w hw,
+    rw [linear_map.add_apply, hφ w hw, hψ w hw, add_zero] },
+  smul_mem' := by { intros c φ hφ w hw,
+    rw [linear_map.smul_apply, hφ w hw, smul_zero] } }
+
+@[simp] lemma mem_dual_annihilator (φ : module.dual R M) :
+  φ ∈ W.dual_annihilator ↔ ∀ w ∈ W, φ w = 0 := iff.rfl
+
+/-- The `dual_restrict` of a submodule `W` of `M` is the linear map from the
+  dual of `M` to the dual of `W` such that the domain of each linear map is
+  restricted to `W`. -/
+def dual_restrict (W : submodule R M) :
+  module.dual R M →ₗ[R] module.dual R W :=
+linear_map.dom_restrict' W
+
+@[simp] lemma dual_restrict_apply
+  (W : submodule R M) (φ : module.dual R M) (x : W) :
+  W.dual_restrict φ x = φ x.1 := rfl
+
+lemma dual_restrict_ker_eq_dual_anihilator
+  (W : submodule R M) : W.dual_restrict.ker = W.dual_annihilator :=
+begin
+  ext φ, split; intro hφ,
+  { intros w hw,
+    rw linear_map.mem_ker at hφ,
+    rw [← W.dual_restrict_apply φ ⟨w, hw⟩, hφ], refl },
+  { ext, exact hφ x.1 x.2 }
+end
+
+end submodule
+
+namespace subspace
+
+open submodule linear_map
+
+universes u v w
+
+-- We work in vector spaces because `exists_is_compl` only hold for vector spaces
+variables {K : Type u} {V : Type v} [field K] [add_comm_group V] [vector_space K V]
+
+/-- Given a subspace `W` of `V` and an element of its dual `φ`, `dual_lift W φ` is
+  the natural extenstion of `φ` to an element of the dual of `V`. -/
+noncomputable def dual_lift
+  (W : subspace K V) (φ : module.dual K W) : module.dual K V :=
+let h := classical.indefinite_description _ W.exists_is_compl in of_is_compl h.2 φ 0
```

Comment:
Is there a linear equiv hiding here between `module.dual K W` and `module.dual K V`, whose inverse is `dual_restrict`?

Reply:
Surely that is not true. The dual of W has the same dimension as W while the dual of V has the same dimension as V, so if the two duals are equiv then every subspace has the same dimension as the whole space.
`dual_restrict` is the left inverse of `dual_lift` however which I will add.

Reply:
I've added `dual_lift_left_inverse` which show that dual_restrict` is the left inverse of `dual_lift`